### ISSUE TITLE
guys, really need some help about multi-devices(multi-GPU)training

### ISSUE THREAD

**Issue by 652994331** [horovod/horovod#1397]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet) tensorflow
2. Framework version: 1.14.0
3. Horovod version: 0.18.0
4. MPI version:4.0.1
5. CUDA version:10
6. NCCL version:2.4.8 for cuda 10
7. Python version:3.6.9
8. OS and version: centos7
9. GCC version:>4.8.5

Hi, guys, i have two devices with 4 gpus on each. i already set up the same environment for each device. And after i build a no password ssh between these two devices, i can simply start a 8 gpu training with example file tensorflow_mist.py   using: horovodrun -np 8 -H server1:4, server2:4 python3 mycode.py. So then, i got my real training data and start a bert-pretraining. Then some wrong here. at first , i started the 8 devices training in same way, and it's stop with no error.(no idea why, just stop there.) then, I start 4 gpu training on each device, it;'s ok, training started. But after i tried 8 devices again, it said:
![image](https://user-images.githubusercontent.com/51428350/64831234-9a736380-d589-11e9-817f-d8fc60d124db.png)

someone please help me out, thanks

**Comment by 652994331** [comment]
@tgaddair thanks so much for your help. I found the solution yesterday, after i use mpirun with parameters instead of horovodrun, it  worked. still dont know why this happened, but anyway, it works now. Then here I encountered a problem, i am using two devices with 4 gpus on each(8 in total) but the train speed is not as fast as what i  was expected. after i use nvidia-smi to check the status of the GPUS. I found a problem, below you can see the status of my gpus
![image](https://user-images.githubusercontent.com/51428350/64883194-aad02080-d613-11e9-9356-1b46ec355643.png)

![image](https://user-images.githubusercontent.com/51428350/64883321-f4207000-d613-11e9-95ed-21f0b39e019e.png)
the usage of GPU keep changing from 99% - lower than 5%, it seems it cant maintain a high usage, and that's why we have a relative low training speed, could you please help me with this question? thanks

**Comment by 652994331** [comment]
![image](https://user-images.githubusercontent.com/51428350/64883762-fafbb280-d614-11e9-90ff-b9b187a4d3a1.png)
sorry for the duplicate images, there is the low status screen shot. btw, i am using watch -n 2 nvidia-smi.  the normal status(99%) even cant maintain 2s.

**Comment by tgaddair** [comment]
Hey @652994331, can you provide the `mpirun` that ran successfully and the `horovodrun` command that did not?  That information will be very useful for us to fix issues like this with `horovodrun` that may arise for other users.

For the performance issue, have you tried running `tensorflow_synthetic_benchmark.py`?  Do you get similar results?  What is the scaling efficiency as you go from 1 GPU to 2, etc.?  If the results are good with `tensorflow_synthetic_benchmark.py`, then the issue is likely with your script, which we can dive into more.  If the results are still bad, then it's likely an environment issue, and I'd suggest generating a [timeline](https://github.com/horovod/horovod/blob/master/docs/timeline.rst) that will help us debug the issue further.  You can find instructions on generating a timeline with mpirun [here](https://github.com/horovod/horovod/blob/master/docs/mpirun.rst).

**Comment by 652994331** [comment]
mpirun -np  8 -H 10.252.192.90:4,10.252.192.89:4 -bind-to none  -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python3 run_pretraining.py   --input_file=./tf_examples.tfrecord*          --output_dir=./pretrain_jd/pretraining_output_jd \    --do_train=True      --do_eval=True     --bert_config_file=./bert_config.json    --init_checkpoint=./model_new.ckpt          --train_batch_size=32        --max_seq_length=128    --max_predictions_per_seq=20          --num_train_steps=1000000          --num_warmup_steps=10000          --learning_rate=1e-5    --use_fp16  --use_xla --horovod 
**above is the right one(mpirun)**

 horovodrun -np 8 -H 10.252.192.90:4 10.252.192.89:4 python3 run_pretraining.py   --input_file=./tf_examples.tfrecord*          --output_dir=./pretrain_jd/pretraining_output_jd \    --do_train=True      --do_eval=True     --bert_config_file=./bert_config.json    --init_checkpoint=./chinese_L-12_H-768_A-12/bert_model.ckpt          --train_batch_size=32        --max_seq_length=128    --max_predictions_per_seq=20          --num_train_steps=1000000          --num_warmup_steps=10000          --learning_rate=1e-5    --use_fp16  --horovod 
**above is the original wrong one(but i  use it run horovod/examples/tneosorflow_mnist.py, it works!)**
However, when i use horovodrun to run tensorflow_mnist.py. the GPU usage is not very high, the scrip i used was the NVIDIA-DeeplearningExamples -> run_pretraining.py(bert pretrain), it actually from there i came to horovod, so i think the scopt might not be the problem.

Thanks,

**Comment by 652994331** [comment]
@tgaddair  i just finished running tensorflow_synthhetic_benchmark.py, the usage of the gpus around 60%, but no huge fall (fall to 3%-4%). So i guess, that means, the environment is ok(original we cant get 99% for this benchmark.py) and the problem should be my own script. what I was thinking is, the usage age, as you can see from my previous screen shot, it can reach 99%-100%，the problem is if we use watch nvidia-smi every 2 s, we can see the low speed is because it cant "maintain" a relative high level, so what i was concerned is, it use CPU for some IO operation, and what i want to mention is that, every tfrecord of my data is very large, if i use wc -c, for each tf_record, the size is around 5745614848 and i have 40 tf_records. So maybe, the original,some parts from original scrip(nvidia-pretrain) contains IO operation with CPU. If that's right, the CPU Operation might be very heavy, that might cause GPU usage fall down dramatically.

**Comment by tgaddair** [comment]
Hey @652994331, I noticed there's no comma in the `horovodrun` script between hosts:

```
horovodrun -np 8 -H 10.252.192.90:4 10.252.192.89:4
```

Assuming that's not a typo, that could be the cause of the issue.  Correct way is identical to how it's done in `mpirun`:

```
horovodrun -np 8 -H 10.252.192.90:4,10.252.192.89:4
```

As you said, your training script may be I/O bound.  I would suggest creating a timeline to debug it further.  Is it possible to decrease record or batch size to see if that might help speed things up?

**Comment by 652994331** [comment]
thanks，for the comma, it's typo. For the tfrecord size, is there any way we can split it? Cuz it's in data format and a txt file. We actually don't want to decrease batch size considering the performance. Thanks for your help. Btw for the time line, after I read the turorial of horovod I only found timeline for horovodrun command. So, for according to my condition(I used mpirun),how can I make a timeline.

发自我的iPhone


------------------ Original ------------------

**Comment by tgaddair** [comment]
For the timeline you use:

```
HOROVOD_TIMELINE=/path/to/timeline.json mpirun ...
```

**Comment by 652994331** [comment]
@tgaddair  when i used this command, something wrong showed up:
![image](https://user-images.githubusercontent.com/51428350/64995818-1ecb2c80-d891-11e9-97fe-41ad33bbdba3.png)
my command is like this:
HOROVOD_TIMELINE=/path/to/timeline.json mpirun -np  8 -H 10.252.192.90:4,10.252.192.89:4 -bind-to none  -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python3 run_pretraining.py   --input_file=./tf_examples.tfrecord*          --output_dir=./pretrain_jd/pretraining_output_jd \    --do_train=True      --do_eval=True     --bert_config_file=./bert_config.json    --init_checkpoint=./model_new.ckpt          --train_batch_size=64        --max_seq_length=128    --max_predictions_per_seq=20          --num_train_steps=1000000          --num_warmup_steps=10000          --learning_rate=1e-5    --use_fp16   --horovod

**Comment by tgaddair** [comment]
Hey @652994331, my mistake on the args, try this:

```
mpirun -np 8 -H 10.252.192.90:4,10.252.192.89:4 -bind-to none -map-by slot -x HOROVOD_TIMELINE=/path/to/timeline.json ...
```

Basically, from `mpirun` the environment variable needs to be provided by `-x`.

**Comment by 652994331** [comment]
it works, thanks so much. the right command line is like below. Btw, how can i collect this timeline, i mean i should collect timeline.json now or wait the training process for a while. Until now, i haven't seen this  /timeline/timeline.json in the current folder
mpirun -np  8 -H 10.252.1990:4,10.252.192.89:4 -bind-to none  -map-by slot -x HOROVOD_TIMELINE=/timeline/timeline.json -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH -mca pml ob1 -mca btl ^openib python3 run_pretraining.py   --input_file=./tf_examples.tfrecord*          --output_dir=./pretrain_jd/pretraining_output_jd \    --do_train=True      --do_eval=True     --bert_config_file=./bert_config.json    --init_checkpoint=./model_new.ckpt          --train_batch_size=32        --max_seq_length=128    --max_predictions_per_seq=20          --num_train_steps=1000000          --num_warmup_steps=10000          --learning_rate=1e-5    --use_fp16   --horovod

**Comment by tgaddair** [comment]
`/timeline/timeline.json` is an absolute path, so if you want it to be written to the current folder, you would need to use a relative path: `HOROVOD_TIMELINE=timeline.json`.

I would suggest letting it run through a few batches to smooth out any noise.  Be careful that these files can grow very large, so I wouldn't run it for too long.

**Comment by 652994331** [comment]
here are some screen shots. looks a little t bit weird..
![image](https://user-images.githubusercontent.com/51428350/65083252-e7c24d00-d95c-11e9-9a81-b980918f58e5.png)
![image](https://user-images.githubusercontent.com/51428350/65083276-fa3c8680-d95c-11e9-9747-a8242d06eaec.png)
![image](https://user-images.githubusercontent.com/51428350/65083316-13ddce00-d95d-11e9-8b7e-4d8f62163c2e.png)
![image](https://user-images.githubusercontent.com/51428350/65083339-25bf7100-d95d-11e9-9dd1-1093c0553a06.png)

**Comment by 652994331** [comment]
@tgaddair  Hi, tgaddair, did you receive my screen shot. still in trouble here, please help, thanks!

**Comment by 652994331** [comment]
this is my timeline.json. please feel free to check,thanks
[timeline.zip](https://github.com/horovod/horovod/files/3632643/timeline.zip)

**Comment by tgaddair** [comment]
Hey @652994331, could you try uploading the timeline json again?  The file appears to be empty.

**Comment by 652994331** [comment]
@got it , tgaddair, thanks for your reply. Btw, let me share some things i did yesterday. I found the nccl I used is for cuda-9.0.(i installed 10.0 myself, but later someone used these machines). So I installed nccl for cuda_10(did not uninstall old nccl, I think it will overwrite). now, it seems , the huge wave of GPU usage disappears.
then, i read an article about the training speed of horovod, it seems i should not use --use_fp16 cuz both my machines are p40. So, i remove that command. 
after all these, now i feel the speed is better than before, but still not fast, i mean, i used 8 gpus, so at least i should get 4X speed than a single gpu, even if my dataset is large. 
So, I am generating this new timeline.json for you, soon i will upload.please help me to have a check later, thanks so much.

**Comment by 652994331** [comment]
@tgaddair got it , tgaddair, thanks for your reply. Btw, let me share some things i did yesterday. I found the nccl I used is for cuda-9.0.(i installed 10.0 myself, but later someone used these machines). So I installed nccl for cuda_10(did not uninstall old nccl, I think it will overwrite). now, it seems , the huge wave of GPU usage disappears.
then, i read an article about the training speed of horovod, it seems i should not use --use_fp16 cuz both my machines are p40. So, i remove that command.
after all these, now i feel the speed is better than before, but still not fast, i mean, i used 8 gpus, so at least i should get 4X speed than a single gpu, even if my dataset is large.
So, I am generating this new timeline.json for you, soon i will upload.please help me to have a check later, thanks so much.

**Comment by tgaddair** [comment]
Thanks for the update.  Before you were getting around 60% usage on the synthetic benchmark.  Have the numbers improved at all since your changes?

**Comment by 652994331** [comment]
around 100% and here is the screen shot
![image](https://user-images.githubusercontent.com/51428350/65362428-2bff5880-dbbc-11e9-9ba7-f1b087fc8f0d.png)

**Comment by 652994331** [comment]
@tgaddair  hope we can find out the reason, gl!

[timeline.json.zip](https://github.com/horovod/horovod/files/3638094/timeline.json.zip)

**Comment by tgaddair** [comment]
Hey @652994331, here's what I'm noticing in the timeline.  Everything looks fine for the most part, but there are these periodic long pauses of negotiation:

<img width="1386" alt="Screen Shot 2019-09-23 at 8 37 34 AM" src="https://user-images.githubusercontent.com/1742912/65440334-6dbb1980-dddd-11e9-8778-07e23a7e60ef.png">

That would suggest that one of the ranks is doing something else while the others are trying to aggregate their gradients.  For example, the first rank might be doing validation.

It would be tough to say more without seeing the code, but my suspicion is that at the moment one of the ranks is being given additional work to do that's causing the other ranks to have to throttle their performance.

**Comment by 652994331** [comment]
Thanks so much for your reply. I guess it's because each small piece of my data is pretty big, (around 1 million rows). right now the useage is good(around 100%) after I installed right NCCl. Btw, for the code, it's official nvidia deeplearning-&gt;Bert pretraining using horovod from github. Thanks again.

发自我的iPhone


------------------ Original ------------------

**Comment by 652994331** [comment]
Hi,@tgaddair, it's me. it's been a longtime, we finally finished pretrain 200 0000 steps of bert. during this period, what I noticed is, although I made it run using 100% gpu usage with your help. the Pwr:Usage/Cap and memory useage of my gpus afre still not good. what I am thinking right now is: maybe we can make it better through improving them. /Users/qizhou.huang/Desktop/Screen Shot 2019-10-14 at 3.04.19 PM.png

/Users/qizhou.huang/Desktop/Screen Shot 2019-10-14 at 3.04.19 PM.png

**Comment by 652994331** [comment]
as you can see , it only 150w/250w ,16691/22919
/Users/qizhou.huang/Desktop/Screen Shot 2019-10-14 at 3.04.19 PM.png

**Comment by MuhammadShifa** [comment]
Thanks both of you for their discussion on the topic. I have 1 question related to this, I don't know how to connect multiple machine having 4 gpus each for training keras model? How to configure cluster of computer for Horovod? Thanks in advance
